helix-term/src/commands.rs

```diff
@@ -3061,6 +3062,123 @@ fn right_bracket_mode(cx: &mut Context) {
     })
 }
 
+fn surround(cx: &mut Context) {
+    let count = cx.count;
+    cx.on_next_key(move |cx, event| {
+        if let KeyEvent {
+            code: KeyCode::Char(ch),
+            ..
+        } = event
+        {
+            // FIXME: count gets reset because of cx.on_next_key()
+            cx.count = count;
+            match ch {
+                'm' => match_brackets(cx),
+                'a' => surround_add(cx),
+                'r' => surround_replace(cx),
+                'd' => {
+                    surround_delete(cx);
+                    let (view, doc) = current!(cx.editor);
+                }
+                _ => (),
+            }
+        }
+    })
+}
+
+use helix_core::surround;
+
+fn surround_add(cx: &mut Context) {
+    cx.on_next_key(move |cx, event| {
+        if let KeyEvent {
+            code: KeyCode::Char(ch),
+            ..
+        } = event
+        {
+            let (view, doc) = current!(cx.editor);
+            let text = doc.text().slice(..);
+            let selection = doc.selection(view.id);
+            let (open, close) = surround::get_pair(ch);
+
+            let mut changes = Vec::new();
+            for (i, range) in selection.iter().enumerate() {
+                let (from, to) = (range.from(), range.to() + 1);
```

Comment:
Need to clamp the range to `max_to` here, like elsewhere:

```
let max_to = doc.text().len_chars().saturating_sub(1);
let to = std::cmp::min(range.to() + 1, max_to)
```

This is a bit different now on master with the CRLF fixes, you might want to do `get_line_ending(text).len_chars()`

Reply:
Clamped with `get_line_ending` 👍🏾